# Copyright 2025 Google LLC
#
# Licensed under the Apache License, Version 2.0 (the "License");
# you may not use this file except in compliance with the License.
# You may obtain a copy of the License at
#
#     http://www.apache.org/licenses/LICENSE-2.0
#
# Unless required by applicable law or agreed to in writing, software
# distributed under the License is distributed on an "AS IS" BASIS,
# WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
# See the License for the specific language governing permissions and
# limitations under the License.


# This file is automatically generated by CrossSync. Do not edit manually.

import pytest
from grpc import RpcError
from google.cloud.bigtable.data._cross_sync import CrossSync

try:
    from unittest import mock
except ImportError:
    import mock
from google.cloud.bigtable.data._sync_autogen.metrics_interceptor import (
    BigtableMetricsInterceptor,
)


def _make_mock_stream_call(values, exc=None):
    """Create a mock call object that can be used for streaming calls"""
    call = CrossSync._Sync_Impl.Mock()

    def gen():
        for val in values:
            yield val
        if exc:
            raise exc

    call.__iter__ = mock.Mock(return_value=gen())
    return call


class TestMetricsInterceptor:
    @staticmethod
    def _get_target_class():
        return BigtableMetricsInterceptor

    def _make_one(self, *args, **kwargs):
        return self._get_target_class()(*args, **kwargs)

    def test_unary_unary_interceptor_success(self):
        """Test that interceptor handles successful unary-unary calls"""
        instance = self._make_one()
        continuation = CrossSync._Sync_Impl.Mock()
        call = continuation.return_value
        details = mock.Mock()
        request = mock.Mock()
        result = instance.intercept_unary_unary(continuation, details, request)
        assert result == call
        continuation.assert_called_once_with(details, request)

    def test_unary_unary_interceptor_failure(self):
        """Test a failed RpcError with metadata"""
        instance = self._make_one()
        exc = RpcError("test")
        continuation = CrossSync._Sync_Impl.Mock(side_effect=exc)
        details = mock.Mock()
        request = mock.Mock()
        with pytest.raises(RpcError) as e:
            instance.intercept_unary_unary(continuation, details, request)
        assert e.value == exc
        continuation.assert_called_once_with(details, request)

    def test_unary_unary_interceptor_failure_generic(self):
        """Test generic exception"""
        instance = self._make_one()
        exc = ValueError("test")
        continuation = CrossSync._Sync_Impl.Mock(side_effect=exc)
        details = mock.Mock()
        request = mock.Mock()
        with pytest.raises(ValueError) as e:
            instance.intercept_unary_unary(continuation, details, request)
        assert e.value == exc
        continuation.assert_called_once_with(details, request)

    def test_unary_stream_interceptor_success(self):
        """Test that interceptor handles successful unary-stream calls"""
        instance = self._make_one()
        continuation = CrossSync._Sync_Impl.Mock(
            return_value=_make_mock_stream_call([1, 2])
        )
        details = mock.Mock()
        request = mock.Mock()
        wrapper = instance.intercept_unary_stream(continuation, details, request)
        results = [val for val in wrapper]
        assert results == [1, 2]
        continuation.assert_called_once_with(details, request)

    def test_unary_stream_interceptor_failure_mid_stream(self):
        """Test that interceptor handles failures mid-stream"""
        instance = self._make_one()
        exc = ValueError("test")
        continuation = CrossSync._Sync_Impl.Mock(
            return_value=_make_mock_stream_call([1], exc=exc)
        )
        details = mock.Mock()
        request = mock.Mock()
        wrapper = instance.intercept_unary_stream(continuation, details, request)
        with pytest.raises(ValueError) as e:
            [val for val in wrapper]
        assert e.value == exc
        continuation.assert_called_once_with(details, request)

    def test_unary_stream_interceptor_failure_start_stream(self):
        """Test that interceptor handles failures at start of stream with RpcError with metadata"""
        instance = self._make_one()
        exc = RpcError("test")
        continuation = CrossSync._Sync_Impl.Mock()
        continuation.side_effect = exc
        details = mock.Mock()
        request = mock.Mock()
        with pytest.raises(RpcError) as e:
            instance.intercept_unary_stream(continuation, details, request)
        assert e.value == exc
        continuation.assert_called_once_with(details, request)

    def test_unary_stream_interceptor_failure_start_stream_generic(self):
        """Test that interceptor handles failures at start of stream with generic exception"""
        instance = self._make_one()
        exc = ValueError("test")
        continuation = CrossSync._Sync_Impl.Mock()
        continuation.side_effect = exc
        details = mock.Mock()
        request = mock.Mock()
        with pytest.raises(ValueError) as e:
            instance.intercept_unary_stream(continuation, details, request)
        assert e.value == exc
        continuation.assert_called_once_with(details, request)
